### ISSUE TITLE
[V3.0.0][consumer] query matching with some special characters doesn't work

### ISSUE THREAD

**Issue by dpakach** [pact-foundation/pact-js#589]
### Software versions

_Please provide at least OS and version of pact-js_

- **OS**: Ubuntu 20.10
- **Consumer Pact library**:" @pact-foundation/pact": "^10.0.0-beta.22"

### Issue Checklist

Please confirm the following:

- [ ] I have upgraded to the latest
- [ ] I have the read the FAQs in the Readme
- [ ] I have triple checked, that there are **no unhandled promises** in my code
- [ ] I have set my log level to debug and attached a log file showing the complete request/response cycle
- [ ] For bonus points and virtual high fives, I have created a reproduceable git repository (see below) to illustrate the problem

### Description
Adding a interaction with query parameters containing some special foreign character doesn't match

### Expected behaviour
interaction with work with query parameters encoded properly

### Actual behaviour
some characters don't work and fail

### Steps to reproduce
https://github.com/dpakach/pact-js/pull/1/files

### Relevant log files
```
        Mock server failed with the following mismatches: 
          1) QueryMismatch Expected '文件.txt' to be equal to 'æä»¶.txt'
      at node_modules/@pact-foundation/src/v3/pact.ts:219:35

**Comment by individual-it** [comment]
adding `encodeURIComponent('文件.txt')` on both sides (consumer & test) makes the test pass
but then the query is double encoded
the consumer sends `%25E6%2596%2587%25E4%25BB%25B6`  that is the encoded version of `%E6%96%87%E4%BB%B6` what again is the encoded version of `文件`

**Comment by individual-it** [comment]
to me it looks like the problem is in https://github.com/pact-foundation/pact-ruby-standalone/
it seems to get the value from the URL, decode it and then when comparing not using the UTF value but some other encoding

**Comment by TimothyJones** [comment]
@individual-it I don't think it'll be in the ruby standalone, as this is using the rust bindings.

@uglyog, any insight as to what might be happening here?

**Comment by individual-it** [comment]
here the log output of LOG_LEVEL=debug
```
RUNS  consumer/transaction-service.test.js
[2021-01-27T04:35:24Z DEBUG pact_js_v3] Initialising Pact native library version 0.0.6
[2021-01-27T04:35:24Z DEBUG pact_js_v3] Request = Request ( method: GET, path: /accounts/search/findOneByAccountNumberId, query: Some({"accountNumber": ["100"], "anotherValue": ["文件.txt"]}), headers: Some({"Accept": ["application/hal+json; charset=utf-8"]}), body: Missing )
[2021-01-27T04:35:24Z DEBUG pact_js_v3] Request matching rules = MatchingRules { rules: {"query": MatchingRuleCategory { name: "query", rules: {"accountNumber": RuleList { rules: [Type], rule_logic: And }} }} }
[2021-01-27T04:35:24Z DEBUG pact_js_v3] Request generators = Generators { categories: {QUERY: {"accountNumber": ProviderStateGenerator("${accountNumber}", None)}} }
[2021-01-27T04:35:24Z DEBUG pact_js_v3] Generated datetime value from 'yyyy-MM-dd'T'HH:mm:ss.SSSX' -> '2021-01-27T10:20:24.701+05:45'
[2021-01-27T04:35:24Z DEBUG pact_js_v3] Generated datetime value from 'yyyy-MM-dd'T'HH:mm:ss.SSSX' -> '2021-01-27T10:20:24.702+05:45'
[2021-01-27T04:35:24Z DEBUG pact_js_v3] Response = Response ( status: 200, headers: Some({"Content-Type": ["application/hal+json"]}), body: Present(303 bytes, application/json) )
[2021-01-27T04:35:24Z DEBUG pact_js_v3] Response matching rules = MatchingRules { rules: {"body": MatchingRuleCategory { name: "body", rules: {"$._links.account.href": RuleList { rules: [Regex(".*(\\/accounts\\/\\d+)$")], rule_logic: And }, "$.version": RuleList { rules: [Integer], rule_logic: And }, "$.lastModifiedDate": RuleList { rules: [Timestamp("yyyy-MM-dd\'T\'HH:mm:ss.SSSX")], rule_logic: And }, "$.createdDate": RuleList { rules: [Timestamp("yyyy-MM-dd\'T\'HH:mm:ss.SSSX")], rule_logic: And }, "$.accountNumber.id": RuleList { rules: [Type], rule_logic: And }, "$.id": RuleList { rules: [Integer], rule_logic: And }, "$._links.self.href": RuleList { rules: [Regex(".*(\\/accounts\\/\\d+)$")], rule_logic: And }, "$.accountRef": RuleList { rules: [Type], rule_logic: And }, "$.name": RuleList { rules: [Type], rule_logic: And }} }} }
[2021-01-27T04:35:24Z DEBUG pact_js_v3] Response generators = Generators { categories: {BODY: {"$.accountNumber.id": ProviderStateGenerator("${accountNumber}", None), "$.lastModifiedDate": DateTime(Some("[2021-01-27T04:35:24.546Z] DEBUG: pact@10.0.0-beta.25/33783 on artur-OptiPlex-3050: Initialised native library 0.0.6:ss.SSSX"))}} }

 RUNS  consumer/transaction-service.test.js
[2021-01-27T04:35:24Z DEBUG pact_mock_server::hyper_server] Creating pact request from hyper request
[2021-01-27T04:35:24Z INFO  pact_mock_server::hyper_server] Received request Request ( method: GET, path: /accounts/search/findOneByAccountNumberId, query: Some({"anotherValue": ["æ\u{96}\u{87}ä»¶.txt"], "accountNumber": ["100"]}), headers: Some({"origin": ["http://localhost"], "host": ["127.0.0.1:1234"], "accept-encoding": ["gzip", "deflate"], "accept-language": ["en"], "connection": ["keep-alive"], "user-agent": ["Mozilla/5.0 (linux) AppleWebKit/537.36 (KHTML", "like Gecko) jsdom/16.4.0"], "referer": ["http://localhost/"], "accept": ["application/hal+json; charset=utf-8"]}), body: Empty )
[2021-01-27T04:35:24Z INFO  pact_matching] comparing to expected Request ( method: GET, path: /accounts/search/findOneByAccountNumberId, query: Some({"accountNumber": ["100"], "anotherValue": ["文件.txt"]}), headers: Some({"Accept": ["application/hal+json; charset=utf-8"]}), body: Missing )
[2021-01-27T04:35:24Z DEBUG pact_matching]      body: ''
[2021-01-27T04:35:24Z DEBUG pact_matching]      matching_rules: MatchingRules { rules: {"query": MatchingRuleCategory { name: "query", rules: {"accountNumber": RuleList { rules: [Type], rule_logic: And }} }} }
[2021-01-27T04:35:24Z DEBUG pact_matching]      generators: Generators { categories: {QUERY: {"accountNumber": ProviderStateGenerator("${accountNumber}", None)}} }
[2021-01-27T04:35:24Z DEBUG pact_matching::matchers] String -> String: comparing '/accounts/search/findOneByAccountNumberId' to '/accounts/search/findOneByAccountNumberId' using Equality
[2021-01-27T04:35:24Z DEBUG pact_matching] expected content type = '*/*', actual content type = '*/*'
[2021-01-27T04:35:24Z DEBUG pact_matching] content type header matcher = 'None'
[2021-01-27T04:35:24Z DEBUG pact_matching::matchers] String -> String: comparing '100' to '100' using Type
[2021-01-27T04:35:24Z DEBUG pact_matching::matchers] String -> String: comparing '文件.txt' to 'æä»¶.txt' using Equality
[2021-01-27T04:35:24Z DEBUG pact_matching] --> Mismatches: [QueryMismatch { parameter: "anotherValue", expected: "文件.txt", actual: "æ\u{96}\u{87}ä»¶.txt", mismatch: "Expected \'文件.txt\' to be equal to \'æ\u{96}\u{87}ä»¶.txt\'" }]
 FAIL  consumer/transaction-service.test.jsr::hyper_server] Request did not match: Request did not match - Request ( method: GET, path: /accounts/search/findOneByAccountNumberId, query: Some({"accountNumb  Transaction service - create a new transaction for an account

**Comment by uglyog** [comment]
Fixed in the downstream Rust library. It was not assembling multi-byte UTF-8 encoded characters correctly.

**Comment by individual-it** [comment]
Using Devanagari characters in the query I get a bit of a different error
I hope the fix will be be valid for both
```
[2021-02-10T07:16:39Z DEBUG pact_matching]      matching_rules: MatchingRules { rules: {} }
[2021-02-10T07:16:39Z DEBUG pact_matching]      generators: Generators { categories: {} }
[2021-02-10T07:16:39Z DEBUG pact_matching::matchers] String -> String: comparing '/animals/available' to '/animals/available' using Equality
[2021-02-10T07:16:39Z DEBUG pact_matching] expected content type = '*/*', actual content type = '*/*'
[2021-02-10T07:16:39Z DEBUG pact_matching] content type header matcher = 'None'                         à¤²à¥' using Equality
[2021-02-10T07:16:39Z DEBUG pact_matching::matchers] String -> String: comparing 'Bearer token' to 'Bearer token' using Equality
[2021-02-10T07:16:39Z DEBUG pact_matching] --> Mismatches: [QueryMismatch { parameter: "first_name", expected: "बिल\u{94d}ली", actual: "à¤¬à¤¿à¤²à¥\u{8d}à¤²à¥\u{80}", mismatch: "Expected \'बिल\u{94d}ली\' to be equal to \'à¤¬à¤¿à¤²à¥\u{8d}à¤²à¥\u{80}\'" }]
[2021-02-10T07:16:39Z DEBUG pact_mock_server::hyper_server] Request did not match: Request did not match - Request ( method: GET, path: /animals/available, query: Sà¤²à¥'first_name": ["बिल\u{94d}ली"]}), headers: Some({"Authorization": ["Bearer token"]}), body: Missing )    0) Expected 'बिल्ली' to be equal to 'à¤¬à          1) returns a filtered list of animals (query containing devanagari characters)

  3 passing (60ms)
  1 pending
  1 failing

  1) Pact V3
       when a call to list all animals from the Animal Service is made
         and the user is authenticated
           and there are animals in the database
             returns a filtered list of animals (query containing devanagari characters):
     Error: Test failed for the following reasons:

  Test code failed with an error: Internal Server Error
Error: Internal Server Error
    at Request.callback (/home/artur/www/pact/pact-js/examples/v3/e2e/node_modules/superagent/lib/node/index.js:706:15)
    at /home/artur/www/pact/pact-js/examples/v3/e2e/node_modules/superagent/lib/node/index.js:916:18
    at IncomingMessage.<anonymous> (/home/artur/www/pact/pact-js/examples/v3/e2e/node_modules/superagent/lib/node/parsers/json.js:19:7)
    at IncomingMessage.emit (events.js:326:22)
    at endReadableNT (_stream_readable.js:1241:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)


  Mock server failed with the following mismatches:              à¤²à¥'
      at /home/artur/www/pact/pact-js/examples/v3/e2e/node_modules/@pact-foundation/src/v3/pact.ts:227:35
      at processTicksAndRejections (internal/process/task_queues.js:97:5)

**Comment by individual-it** [comment]
tests pass in #603 I think we can close this issue, or is an update/release needed?

**Comment by uglyog** [comment]
This is fixed with v10.0.0-beta.31

**Comment by individual-it** [comment]
works great, thanks! @dpakach please close this issue, I don't have permissions for that
